File: scripts/generate-app-icons.mjs
#!/usr/bin/env node
/**
 * SyncedIn app icon generator.
 *
 * Renders the master 1024x1024 brand icon as SVG, then uses ImageMagick
 * to rasterize it to every size iOS + Android expect. Drop the output
 * folder contents into:
 *   - ios/App/App/Assets.xcassets/AppIcon.appiconset/   (iOS catalog)
 *   - android/app/src/main/res/mipmap-*                 (Android)
 *
 * Run from project root:
 *   node scripts/generate-app-icons.mjs
 *
 * Requires `convert` (ImageMagick) on PATH.
 */

import { execSync } from "node:child_process";
import { writeFileSync, mkdirSync, existsSync } from "node:fs";
import { join, resolve, dirname } from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = dirname(fileURLToPath(import.meta.url));
const OUT_DIR = resolve(__dirname, "..", "public", "app-icons");
mkdirSync(OUT_DIR, { recursive: true });

/**
 * Master 1024x1024 icon. Brand mark = twin-pair interlocked circles on
 * the SyncedIn gradient. White-hot inner glow conveys "two clones
 * about to sync". Rounded square corner radius matches iOS native icon
 * grid (180/1024 ≈ 17.6% — Apple applies the squircle mask, but the
 * App Store still expects a square master, so we render a square here
 * and the OS handles the corners).
 */
function masterSvg() {
  return `<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 1024 1024">
  <defs>
    <linearGradient id="bg" x1="0" y1="0" x2="1" y2="1">
      <stop offset="0%" stop-color="#05021f"/>
      <stop offset="35%" stop-color="#1e3aff"/>
      <stop offset="65%" stop-color="#6b2dc9"/>
      <stop offset="100%" stop-color="#d83bff"/>
    </linearGradient>
    <radialGradient id="glow" cx="0.5" cy="0.5" r="0.55">
      <stop offset="0%" stop-color="rgba(255,255,255,0.55)"/>
      <stop offset="60%" stop-color="rgba(255,255,255,0.05)"/>
      <stop offset="100%" stop-color="rgba(255,255,255,0)"/>
    </radialGradient>
    <linearGradient id="ringLeft" x1="0" y1="0" x2="1" y2="1">
      <stop offset="0%" stop-color="#ffffff"/>
      <stop offset="100%" stop-color="#a8b6ff"/>
    </linearGradient>
    <linearGradient id="ringRight" x1="0" y1="0" x2="1" y2="1">
      <stop offset="0%" stop-color="#ff9cff"/>
      <stop offset="100%" stop-color="#ffffff"/>
    </linearGradient>
    <filter id="softGlow" x="-20%" y="-20%" width="140%" height="140%">
      <feGaussianBlur in="SourceAlpha" stdDeviation="12"/>
      <feOffset dx="0" dy="6"/>
      <feComponentTransfer><feFuncA type="linear" slope="0.55"/></feComponentTransfer>
      <feMerge><feMergeNode/><feMergeNode in="SourceGraphic"/></feMerge>
    </filter>
  </defs>
  <rect width="1024" height="1024" fill="url(#bg)"/>
  <rect width="1024" height="1024" fill="url(#glow)"/>
  <g filter="url(#softGlow)">
    <circle cx="400" cy="512" r="180" fill="none" stroke="url(#ringLeft)" stroke-width="48"/>
    <circle cx="624" cy="512" r="180" fill="none" stroke="url(#ringRight)" stroke-width="48"/>
    <circle cx="512" cy="512" r="32" fill="#ffffff"/>
  </g>
</svg>`;
}

const masterSvgPath = join(OUT_DIR, "icon.master.svg");
writeFileSync(masterSvgPath, masterSvg());

// iOS App Icon catalog: every size the Asset Catalog wants. Sizes are
// expressed as pixels (already 1x/2x/3x flattened) since that's what
// `convert` outputs and what Xcode happily ingests via drag-drop.
const IOS_SIZES = [
  { name: "Icon-40", px: 40 },
  { name: "Icon-58", px: 58 },
  { name: "Icon-60", px: 60 },
  { name: "Icon-76", px: 76 },
  { name: "Icon-80", px: 80 },
  { name: "Icon-87", px: 87 },
  { name: "Icon-120", px: 120 },
  { name: "Icon-152", px: 152 },
  { name: "Icon-167", px: 167 },
  { name: "Icon-180", px: 180 },
  { name: "Icon-1024", px: 1024 } // App Store marketing icon
];

const ANDROID_SIZES = [
  { name: "ic_launcher_48", px: 48 }, // mdpi
  { name: "ic_launcher_72", px: 72 }, // hdpi
  { name: "ic_launcher_96", px: 96 }, // xhdpi
  { name: "ic_launcher_144", px: 144 }, // xxhdpi
  { name: "ic_launcher_192", px: 192 }, // xxxhdpi
  { name: "ic_launcher_512", px: 512 } // Play Store listing icon
];

function render(target, sizePx) {
  const out = join(OUT_DIR, `${target}.png`);
  execSync(
    `convert -background none -density 384 -resize ${sizePx}x${sizePx} "${masterSvgPath}" "${out}"`,
    { stdio: "inherit" }
  );
  return out;
}

console.log(`Rendering iOS icons → ${OUT_DIR}/ios/`);
mkdirSync(join(OUT_DIR, "ios"), { recursive: true });
for (const s of IOS_SIZES) {
  const out = join("ios", s.name);
  render(out, s.px);
}

console.log(`Rendering Android icons → ${OUT_DIR}/android/`);
mkdirSync(join(OUT_DIR, "android"), { recursive: true });
for (const s of ANDROID_SIZES) {
  const out = join("android", s.name);
  render(out, s.px);
}

// Android adaptive icon = a foreground SVG (transparent bg) + a background
// SVG (solid color/gradient). Android Studio composites them on-device.
const adaptiveFg = `<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 108 108">
  <defs>
    <linearGradient id="ringLeft" x1="0" y1="0" x2="1" y2="1">
      <stop offset="0%" stop-color="#ffffff"/>
      <stop offset="100%" stop-color="#a8b6ff"/>
    </linearGradient>
    <linearGradient id="ringRight" x1="0" y1="0" x2="1" y2="1">
      <stop offset="0%" stop-color="#ff9cff"/>
      <stop offset="100%" stop-color="#ffffff"/>
    </linearGradient>
  </defs>
  <g transform="translate(54 54)">
    <circle cx="-12" cy="0" r="14" fill="none" stroke="url(#ringLeft)" stroke-width="4"/>
    <circle cx="12" cy="0" r="14" fill="none" stroke="url(#ringRight)" stroke-width="4"/>
    <circle cx="0" cy="0" r="2.5" fill="#ffffff"/>
  </g>
</svg>`;
const adaptiveBg = `<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 108 108">
  <defs>
    <linearGradient id="bg" x1="0" y1="0" x2="1" y2="1">
      <stop offset="0%" stop-color="#05021f"/>
      <stop offset="35%" stop-color="#1e3aff"/>
      <stop offset="65%" stop-color="#6b2dc9"/>
      <stop offset="100%" stop-color="#d83bff"/>
    </linearGradient>
  </defs>
  <rect width="108" height="108" fill="url(#bg)"/>
</svg>`;
writeFileSync(join(OUT_DIR, "android", "ic_launcher_foreground.svg"), adaptiveFg);
writeFileSync(join(OUT_DIR, "android", "ic_launcher_background.svg"), adaptiveBg);

// Render adaptive PNGs too (Android Studio prefers PNG fallbacks).
execSync(
  `convert -background none -density 384 -resize 432x432 "${join(OUT_DIR, "android", "ic_launcher_foreground.svg")}" "${join(OUT_DIR, "android", "ic_launcher_foreground.png")}"`
);
execSync(
  `convert -background none -density 384 -resize 432x432 "${join(OUT_DIR, "android", "ic_launcher_background.svg")}" "${join(OUT_DIR, "android", "ic_launcher_background.png")}"`
);

// Screenshot template — 1290×2796 (iPhone 16 Pro Max safe). Renders the
// status-bar + a centered placeholder with brand chrome. Jack drops
// his own Cowork captures over the placeholder area when filing the
// App Store listing.
const screenshotSvg = `<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 1290 2796">
  <defs>
    <linearGradient id="frame" x1="0" y1="0" x2="0" y2="1">
      <stop offset="0%" stop-color="#0b0b16"/>
      <stop offset="100%" stop-color="#1a0f2a"/>
    </linearGradient>
  </defs>
  <rect width="1290" height="2796" fill="url(#frame)"/>
  <text x="645" y="240" text-anchor="middle" fill="#ffffff" font-family="-apple-system, system-ui, sans-serif" font-weight="800" font-size="92">SyncedIn</text>
  <text x="645" y="310" text-anchor="middle" fill="#a8b6ff" font-family="-apple-system, system-ui, sans-serif" font-weight="500" font-size="38">Twin-to-twin networking</text>
  <rect x="80" y="420" width="1130" height="2000" rx="64" fill="rgba(255,255,255,0.06)" stroke="rgba(255,255,255,0.15)" stroke-width="2"/>
  <text x="645" y="1400" text-anchor="middle" fill="rgba(255,255,255,0.35)" font-family="-apple-system, system-ui, sans-serif" font-weight="500" font-size="36">screenshot goes here · 1130 × 2000</text>
  <text x="645" y="2620" text-anchor="middle" fill="#ffffff" font-family="-apple-system, system-ui, sans-serif" font-weight="700" font-size="58">syncedin.org</text>
</svg>`;
writeFileSync(join(OUT_DIR, "screenshot-template.svg"), screenshotSvg);
execSync(
  `convert -background none -density 192 -resize 1290x2796 "${join(OUT_DIR, "screenshot-template.svg")}" "${join(OUT_DIR, "screenshot-template.png")}"`
);

console.log(`\n✓ Done. Output at: ${OUT_DIR}`);
console.log(`  - ios/Icon-1024.png         → App Store marketing icon`);
console.log(`  - ios/Icon-180.png          → iPhone 60pt @3x`);
console.log(`  - ios/Icon-167.png          → iPad Pro 83.5pt @2x`);
console.log(`  - android/ic_launcher_*.png → Drop into res/mipmap-*`);
console.log(`  - screenshot-template.png   → 6.7-inch iPhone listing template`);

File: scripts/generate-app-icons.py
#!/usr/bin/env python3
"""
SyncedIn app-icon generator (PIL native, no SVG engine required).

Renders a 1024x1024 master icon as a pure PIL image, then resizes to
every iOS + Android size. Drop the resulting `ios/` and `android/`
folders into the respective native projects.

Usage:
    python3 scripts/generate-app-icons.py
"""

import os
from pathlib import Path

from PIL import Image, ImageDraw, ImageFilter

ROOT = Path(__file__).resolve().parent.parent
OUT = ROOT / "public" / "app-icons"
OUT.mkdir(parents=True, exist_ok=True)

MASTER = 1024


def lerp_color(c1, c2, t):
    return tuple(int(c1[i] + (c2[i] - c1[i]) * t) for i in range(3))


def render_gradient_bg(size):
    """Diagonal gradient stops matching the SyncedIn brand palette."""
    img = Image.new("RGB", (size, size), (5, 2, 31))
    pixels = img.load()
    stops = [
        (0.00, (5, 2, 31)),       # #05021f deep
        (0.35, (30, 58, 255)),    # #1e3aff electric blue
        (0.65, (107, 45, 201)),   # #6b2dc9 violet
        (1.00, (216, 59, 255)),   # #d83bff hot magenta
    ]
    diag = (size - 1) * 2
    for y in range(size):
        for x in range(size):
            t = (x + y) / diag
            # Find the active stop segment.
            for i in range(len(stops) - 1):
                t0, c0 = stops[i]
                t1, c1 = stops[i + 1]
                if t0 <= t <= t1:
                    local = (t - t0) / (t1 - t0) if t1 != t0 else 0
                    pixels[x, y] = lerp_color(c0, c1, local)
                    break
            else:
                pixels[x, y] = stops[-1][1]
    return img


def add_radial_glow(img):
    """White-hot center bloom layered over the gradient."""
    size = img.size[0]
    glow = Image.new("RGBA", (size, size), (0, 0, 0, 0))
    g_pix = glow.load()
    cx, cy = size / 2, size / 2
    max_r = size * 0.55
    for y in range(size):
        for x in range(size):
            d = ((x - cx) ** 2 + (y - cy) ** 2) ** 0.5
            t = max(0, 1 - d / max_r)
            # Cubic falloff so the center reads bright but tails off fast.
            a = int(140 * (t ** 2))
            if a > 0:
                g_pix[x, y] = (255, 255, 255, a)
    img = img.convert("RGBA")
    img.alpha_composite(glow)
    return img


def draw_twin_rings(img):
    """Two interlocked rings + center dot. Drawn supersized then
    downsampled for clean anti-aliased edges."""
    size = img.size[0]
    SS = 4  # supersample factor
    layer = Image.new("RGBA", (size * SS, size * SS), (0, 0, 0, 0))
    d = ImageDraw.Draw(layer)

    # Coordinates in master-pixel space, scaled up by SS.
    s = SS
    left_c = (400 * s, 512 * s)
    right_c = (624 * s, 512 * s)
    r = 180 * s
    sw = 48 * s

    # Left ring — cool blue-white
    d.ellipse(
        [left_c[0] - r, left_c[1] - r, left_c[0] + r, left_c[1] + r],
        outline=(220, 232, 255, 255),
        width=sw
    )
    # Right ring — warm magenta-white
    d.ellipse(
        [right_c[0] - r, right_c[1] - r, right_c[0] + r, right_c[1] + r],
        outline=(255, 200, 255, 255),
        width=sw
    )
    # Center dot — pure white
    dot_r = 32 * s
    d.ellipse(
        [size * s / 2 - dot_r, size * s / 2 - dot_r,
         size * s / 2 + dot_r, size * s / 2 + dot_r],
        fill=(255, 255, 255, 255)
    )

    # Downsample for clean edges.
    layer = layer.resize((size, size), Image.LANCZOS)

    # Soft glow behind the rings.
    glow = layer.filter(ImageFilter.GaussianBlur(radius=14))
    img.alpha_composite(glow)
    img.alpha_composite(layer)
    return img


def render_master():
    img = render_gradient_bg(MASTER)
    img = add_radial_glow(img)
    img = draw_twin_rings(img)
    return img.convert("RGB")


def save_set(master, out_dir, sizes):
    out_dir.mkdir(parents=True, exist_ok=True)
    paths = []
    for name, px in sizes:
        resized = master.resize((px, px), Image.LANCZOS)
        p = out_dir / f"{name}.png"
        resized.save(p, "PNG", optimize=True)
        paths.append(p)
        print(f"  ✓ {p.relative_to(ROOT)}  ({px}x{px})")
    return paths


def render_adaptive_foreground(size):
    """Android adaptive icon foreground — transparent bg, rings centered.
    The OS clips this into a 66% safe-zone inside a 108dp canvas, so the
    rings live in the inner ~67% of the rendered image."""
    SS = 4
    img = Image.new("RGBA", (size * SS, size * SS), (0, 0, 0, 0))
    d = ImageDraw.Draw(img)
    cx, cy = size * SS / 2, size * SS / 2
    r = size * SS * 0.18
    sep = size * SS * 0.12
    sw = int(size * SS * 0.05)
    d.ellipse(
        [cx - sep - r, cy - r, cx - sep + r, cy + r],
        outline=(220, 232, 255, 255), width=sw
    )
    d.ellipse(
        [cx + sep - r, cy - r, cx + sep + r, cy + r],
        outline=(255, 200, 255, 255), width=sw
    )
    dot = size * SS * 0.025
    d.ellipse(
        [cx - dot, cy - dot, cx + dot, cy + dot],
        fill=(255, 255, 255, 255)
    )
    return img.resize((size, size), Image.LANCZOS)


def render_adaptive_background(size):
    """Android adaptive icon background — full-bleed brand gradient."""
    return render_gradient_bg(size)


def render_screenshot_template():
    """6.7-inch iPhone screenshot template (1290x2796)."""
    W, H = 1290, 2796
    img = render_gradient_bg(max(W, H))
    img = img.crop((0, 0, W, H))
    d = ImageDraw.Draw(img)
    # Centered placeholder card for Cowork screenshot drop-in.
    card = (80, 420, 1210, 2420)
    d.rounded_rectangle(card, radius=64, fill=(255, 255, 255, 24))
    return img


def main():
    print("Rendering master 1024x1024…")
    master = render_master()
    master.save(OUT / "icon.master.png", "PNG", optimize=True)
    print(f"  ✓ {(OUT / 'icon.master.png').relative_to(ROOT)}")

    ios_sizes = [
        ("Icon-40", 40),
        ("Icon-58", 58),
        ("Icon-60", 60),
        ("Icon-76", 76),
        ("Icon-80", 80),
        ("Icon-87", 87),
        ("Icon-120", 120),
        ("Icon-152", 152),
        ("Icon-167", 167),
        ("Icon-180", 180),
        ("Icon-1024", 1024),
    ]
    print("\niOS icons:")
    save_set(master, OUT / "ios", ios_sizes)

    android_sizes = [
        ("ic_launcher_48", 48),
        ("ic_launcher_72", 72),
        ("ic_launcher_96", 96),
        ("ic_launcher_144", 144),
        ("ic_launcher_192", 192),
        ("ic_launcher_512", 512),
    ]
    print("\nAndroid legacy icons:")
    save_set(master, OUT / "android", android_sizes)

    print("\nAndroid adaptive icon (foreground + background, 432x432):")
    fg = render_adaptive_foreground(432)
    fg.save(OUT / "android" / "ic_launcher_foreground.png", "PNG", optimize=True)
    bg = render_adaptive_background(432)
    bg.save(OUT / "android" / "ic_launcher_background.png", "PNG", optimize=True)
    print("  ✓ android/ic_launcher_foreground.png")
    print("  ✓ android/ic_launcher_background.png")

    print("\nScreenshot template (1290x2796, 6.7-inch iPhone):")
    shot = render_screenshot_template()
    shot.save(OUT / "screenshot-template.png", "PNG", optimize=True)
    print("  ✓ screenshot-template.png")

    print(f"\nAll outputs at: {OUT}")


if __name__ == "__main__":
    main()
